lib/ddtrace/runtime/container.rb

```diff
@@ -0,0 +1,70 @@
+require 'ddtrace/ext/runtime'
+require 'ddtrace/runtime/cgroup'
+
+module Datadog
+  module Runtime
+    # For container environments
+    module Container
+      UUID_PATTERN = '[0-9a-f]{8}[-_]?[0-9a-f]{4}[-_]?[0-9a-f]{4}[-_]?[0-9a-f]{4}[-_]?[0-9a-f]{12}'.freeze
+      CONTAINER_PATTERN = '[0-9a-f]{64}'.freeze
+
+      POD_REGEX = /(pod)?(#{UUID_PATTERN})(?:.slice)?$/
+      CONTAINER_REGEX = /(#{UUID_PATTERN}|#{CONTAINER_PATTERN})(?:.scope)?$/
+
+      Descriptor = Struct.new(
+        :platform,
+        :container_id,
+        :task_uid
+      )
+
+      module_function
+
+      def platform
+        descriptor.platform
+      end
+
+      def container_id
+        descriptor.container_id
+      end
+
+      def task_uid
+        descriptor.task_uid
+      end
+
+      def descriptor
+        @descriptor ||= begin
+          Descriptor.new.tap do |descriptor|
+            begin
+              Cgroup.descriptors.each do |cgroup_descriptor|
+                # Parse container data from cgroup descriptor
+                path = cgroup_descriptor.path
+                next if path.nil?
+
+                # Split path into parts
+                parts = path.split('/')
+                parts.shift # Remove leading empty part
+                next if parts.length < 2
+
+                # Read info from path
+                platform = parts[0]
+                container_id = parts[-1][CONTAINER_REGEX]
+                task_uid = parts[-2][POD_REGEX]
+                # If container ID wasn't found, ignore.
+                # Path might describe a non-container environment.
+                next if container_id.nil?
+
+                descriptor.platform = platform
+                descriptor.container_id = container_id
+                descriptor.task_uid = task_uid
```

Comment:
Do we want to have a `break` here? Once we successfully parse this information out we don't need to hit every other line.

An alternative suggestion could be
```ruby
descriptor.platform = platform if descriptor.platform.nil?
descriptor.container_id = container_id if descriptor.container_id.nil?
descriptor.task_uuid = task_uuid if descriptor.task_uuid.nil?
```

This way if one line has some information but not all, we can grab from a later line.

Reply:
Right so I've waffled on this a bit. I did have a `break`, then removed it because it didn't match the other algo.

There are pros and cons to `break`; pros is its slightly faster (but only for an operation done once), cons are maybe there are consistency issues within the cgroup file itself? I don't know if its possible to see container info striped across the entries, or multiple container IDs/platforms/tasks/etc.... my expectation is these strategies wouldn't handle that well.

To me, seems like nil checking might be the most flexible approach without incurring much additional cost.

Reply:
Looking at this further, I think `break` might be the best approach here.

I wanted to do `nil?` checks, but there's a weakness in regex where you can have partial matches for non-container paths e.g. `/foo/bar` will yield a platform but not a container ID. Obviously `foo` isn't really a platform in this context (because its not a container path), so we should ignore any line that doesn't have a container ID.

Given that rule, then consider the following possible sets of lines present:

1. No lines contain container IDs
2. Some lines have container IDs, all of which match.
3. Some or all lines have container IDs, but at least two of which are different.
4. All lines contain the same container IDs

In scenario 1, we won't want to return anything. In scenario 4, we will return the only value present. Things get weird with the other two; in scenario 2, we'd want the only container ID present. But in scenario 3 (something I've never seen in practice), we have a conflict: given we can only return one, we have to choose one, so why not the first? I just don't think we can handle this scenario well at all anyways.

Therefore, I think `break`ing after reading the first well-formed container line is probably the most efficient approach here.

Reply:
Makes sense.

I agree, while 3 probably won't happen, at least the environments we know about (and care about) are well formed, just returning the first valid should be good enough for now.

Reply:
Changed to reflect plan of action.